File: src/year2024/day7/python/part2.py

# lines = open("../test2.txt").read().strip().split("\n")
lines = open("../prod2.txt").read().strip().split("\n")

def is_line_correct(nums: list[int], idx: int, curr_val: int, target: int) -> bool:
    if idx == len(nums):
        return curr_val == target

    add_res = is_line_correct(nums, idx+1, curr_val+nums[idx], target)
    mul_res = is_line_correct(nums, idx+1, curr_val*nums[idx], target)
    concat_res = is_line_correct(nums, idx+1, int(str(curr_val)+str(nums[idx])), target)
    return add_res or mul_res or concat_res

ans = 0
for line in lines: 
    res, nums = line.split(":")
    res = int(res)
    nums = [int(x) for x in nums.split()]

    if is_line_correct(nums, 0, 0, res):
        ans += res


print(ans)

File: src/year2024/day7/scala/part2.scala


@main def part2(): Long = {
  var lines = scala.io.Source.fromFile("../prod2.txt").getLines().toList
  var ans: Long = 0

  for (line <- lines) {
    val Array(left, right) = line.split(":").map(_.trim)
    val target = left.toLong
    // Splits by ANY amount of whitespace
    val nums = right.split("\\s+").map(_.toInt).toList


    if (Helpers.isLineCorrect(nums, 0, 0, target)) {
      ans += target
    }
  }



  ans
}

object Helpers {
  def isLineCorrect(nums: List[Int], idx: Int, currVal: Long, target: Long): Boolean = {
    if (idx == nums.length) {
      return currVal == target

    }

    var addRes = isLineCorrect(nums, idx+1, currVal+nums(idx), target)
    var mulRes = isLineCorrect(nums, idx+1, currVal*nums(idx), target)
    var concatRes = isLineCorrect(nums, idx+1, (currVal.toString + nums(idx).toString).toLong, target)

    return addRes || mulRes || concatRes
  }
}
